lib/zzhc/Kconfig

```diff
@@ -0,0 +1,41 @@
+#
+# Copyright (c) 2018 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+menu "Self-Registration (Zi ZHu Ce)"
+
+config ZZHC
+	bool "Self-Registration Daemon"
+	select AT_CMD
+	select AT_CMD_PARSER
+	select AT_NOTIF
+	select BASE64
+	select CJSON_LIB
+	select NEWLIB_LIBC
+	select FLASH
+	select FLASH_PAGE_LAYOUT
+	select FLASH_MAP
+	select MPU_ALLOW_FLASH_WRITE
+	select SETTINGS
```

Comment:
We need to hide this option from unrelated platforms. E.g.

depends on BSD_LIB

or similar.

Reply:
The original idea was to have only one (or a few) options to enable the module.

As you can see, having 12 options in order to enable only one library is prone to error, and can be frustrating to users of the library, including customers.

How about this way?
- keeps selecting AT_CMD, AT_CMD_PARSER, AT_NOTIF, BASE64, CJSON_LIB, NEWLIB_LIBC
- depends on SETTINGS
- put FLASH, FLASH_PAGE_LAYOUT, FLASH_MAP, MPU_ALLOW_FLASH_WRITE, NVS and SETTINGS_NVS into prj.conf. (This is learned from example\bluetooth\peripheral_hids_keyboard.)

For users of the SDK, simple is the best.

Reply:
I tried the new version. Using the hello_world sample and the nrf9160_pca10090ns board I found that after first enabling SETTINGS, and then ZZHC, i still ended up with a corrupted configuration because AT_CMD depends on BSD_LIBRARY and could therefore not be selected by ZZHC.

What if we change

depends on SETTINGS

to

depends on BSD_LIBRARY
select SETTINGS

then we still only have to enable one option, but we don't get a corrupted configuration. or

depends on TRUSTED_EXECUTION_NONSECURE
select SETTINGS
select BSD_LIBRARY

if you would prefer to not have the user need to enable anything.

Reply:
@SebastianBoe The same applies to nrf_cloud libraries as well. If you need to make it work in hello_world, then you need to have lots of options in prj.conf to work with (because Kconfig does not support deselection and value assignment).

I have changed the Kconfig file so that all direct dependencies are selected:
`depends on TRUSTED_EXECUTION_NONSECURE`
`select BSD_LIBRARY`
`select SETTINGS`
`select AT_CMD`
`select AT_CMD_PARSER`
`select AT_NOTIF`
`select BASE64`
`select CJSON_LIB`
`select NEWLIB_LIBC`

To make it work on zephyr\samples\hello_world, you need the followings in prj.conf:
`CONFIG_ZZHC=y`

`# Required by Settings`
`CONFIG_FLASH=y`
`CONFIG_FLASH_PAGE_LAYOUT=y`
`CONFIG_FLASH_MAP=y`
`CONFIG_MPU_ALLOW_FLASH_WRITE=y`
`CONFIG_NVS=y`
`CONFIG_SETTINGS_NVS_SECTOR_COUNT=6`

`# Requied by k_malloc()`
`CONFIG_HEAP_MEM_POOL_SIZE=4096`

`# Network`
`CONFIG_NETWORKING=y`
`CONFIG_NET_NATIVE=n`
`CONFIG_NET_SOCKETS=y`
`CONFIG_NET_SOCKETS_OFFLOAD=y`

`# LTE link control`
`CONFIG_LTE_LINK_CONTROL=y`
`CONFIG_LTE_AUTO_INIT_AND_CONNECT=y`